radix-engine/src/engine/native_wrapper.rs

```diff
@@ -0,0 +1,459 @@
+use crate::engine::errors::KernelError;
+use crate::engine::*;
+use crate::model::{
+    TransactionProcessorRunInvocation, WorktopAssertContainsAmountInvocation,
+    WorktopAssertContainsInvocation, WorktopAssertContainsNonFungiblesInvocation,
+    WorktopDrainInvocation, WorktopPutInvocation, WorktopTakeAllInvocation,
+    WorktopTakeAmountInvocation, WorktopTakeNonFungiblesInvocation,
+};
+use crate::types::*;
+use scrypto::resource::AuthZoneDrainInvocation;
+
+// TODO: Cleanup
+pub fn parse_and_invoke_native_function<'y, 'a, Y>(
+    native_function: NativeFunction,
+    args: Vec<u8>,
+    system_api: &'y mut Y,
+) -> Result<ScryptoValue, RuntimeError>
+where
+    Y: SystemApi + Invokable<ScryptoInvocation> + InvokableNative<'a>,
+{
+    match native_function {
+        NativeFunction::EpochManager(EpochManagerFunction::Create) => {
+            let invocation: EpochManagerCreateInvocation = scrypto_decode(&args)
+                .map_err(|e| RuntimeError::KernelError(KernelError::DecodeError(e)))?;
+            system_api
+                .invoke(invocation)
+                .map(|a| ScryptoValue::from_typed(&a))
+        }
+        NativeFunction::ResourceManager(ResourceManagerFunction::BurnBucket) => {
+            let invocation: ResourceManagerBurnInvocation = scrypto_decode(&args)
+                .map_err(|e| RuntimeError::KernelError(KernelError::DecodeError(e)))?;
+            system_api
+                .invoke(invocation)
+                .map(|a| ScryptoValue::from_typed(&a))
+        }
+        NativeFunction::ResourceManager(ResourceManagerFunction::Create) => {
+            let invocation: ResourceManagerCreateInvocation = scrypto_decode(&args)
+                .map_err(|e| RuntimeError::KernelError(KernelError::DecodeError(e)))?;
+            system_api
+                .invoke(invocation)
+                .map(|a| ScryptoValue::from_typed(&a))
+        }
+        NativeFunction::TransactionProcessor(TransactionProcessorFunction::Run) => {
+            let invocation: TransactionProcessorRunInvocation = scrypto_decode(&args)
+                .map_err(|e| RuntimeError::KernelError(KernelError::DecodeError(e)))?;
+            system_api
+                .invoke(invocation)
+                .map(|a| ScryptoValue::from_typed(&a))
+        }
+        NativeFunction::Package(PackageFunction::Publish) => {
+            let invocation: PackagePublishInvocation = scrypto_decode(&args)
+                .map_err(|e| RuntimeError::KernelError(KernelError::DecodeError(e)))?;
+            system_api
+                .invoke(invocation)
+                .map(|a| ScryptoValue::from_typed(&a))
+        }
+    }
+}
+
+// TODO: Cleanup
```

Comment:
Just double check that the clean-up is to decouple `Wasm::NativeInvocation` and `Transaction::NativeInvocation` right? :)

Reply:
yup exactly